internal/collectors/gitlab/repository_collector.go

```diff
@@ -0,0 +1,189 @@
+			repo := r
+			gw.Do(func() {
+				project, _, err := rc.Client.Client().Projects.GetProject(repo.Owner+"%2F"+repo.Name, &gitlab2.GetProjectOptions{})
+				if err != nil {
+					log.Println(err.Error())
+					return
+				}
+
+				rc.extendedCollect(project)
+			})
+
+		}
+		gw.Wait()
+	})
+}
+
+func (rc *repositoryCollector) extendProjectWithProtectedBranches(project gitlab_collected.Repository) gitlab_collected.Repository {
+	var completeProtectedBranches []*gitlab2.ProtectedBranch
+	options := gitlab2.ListProtectedBranchesOptions{}
+
+	err := gitlab.PaginateResults(func(opts *gitlab2.ListOptions) (*gitlab2.Response, error) {
+		projectProtectedBranches, resp, err := rc.Client.Client().ProtectedBranches.ListProtectedBranches(project.ID, &options)
+		if err != nil {
+			return nil, err
+		}
+		completeProtectedBranches = append(completeProtectedBranches, projectProtectedBranches...)
+
+		return resp, nil
+	}, (*gitlab2.ListOptions)(&options))
+	if err != nil {
+		log.Printf("failed to list projects %s", err)
+	}
+
+	extendedProject := project
+	extendedProject.ProtectedBranches = completeProtectedBranches
+	return extendedProject
+}
+
+func (rc *repositoryCollector) extendProjectWithMembers(project gitlab_collected.Repository) gitlab_collected.Repository {
+	var completeMembersList []*gitlab2.ProjectMember
+	options := &gitlab2.ListProjectMembersOptions{}
+
+	err := gitlab.PaginateResults(func(opts *gitlab2.ListOptions) (*gitlab2.Response, error) {
+		projectMembers, resp, err := rc.Client.Client().ProjectMembers.ListAllProjectMembers(project.ID, options)
+		if err != nil {
+			return nil, err
+		}
+		completeMembersList = append(completeMembersList, projectMembers...)
+		return resp, nil
+	}, &options.ListOptions)
+
+	if err != nil {
+		log.Printf("failed to list projects %s", err)
+	}
+
+	extendedProject := project
+	extendedProject.Members = completeMembersList
+	return extendedProject
+}
+
+func (rc *repositoryCollector) collectAll() collectors.SubCollectorChannels {
+	return rc.WrappedCollection(func() {
+
+		var completeProjectsList []*gitlab2.Project
+		maintainerPermissions := gitlab2.MaintainerPermissions
+		options := gitlab2.ListProjectsOptions{MinAccessLevel: &maintainerPermissions}
+
+		organizations, err := rc.Client.Organizations()
+		if err != nil {
+			log.Printf("failed to collect list of orgniazations to get repositories  %s", err)
+			return
+		}
+
+		for _, org := range organizations {
+			err := gitlab.PaginateResults(func(opts *gitlab2.ListOptions) (*gitlab2.Response, error) {
+				repos, resp, err := rc.Client.Client().Groups.ListGroupProjects(org.Name, &gitlab2.ListGroupProjectsOptions{})
+				if err != nil {
+					return nil, err
+				}
+				completeProjectsList = append(completeProjectsList, repos...)
+				return resp, nil
+			}, &options.ListOptions)
+			if err != nil {
+				log.Printf("failed to list projects %s", err)
+			}
+		}
+
+		for _, completeProject := range completeProjectsList {
```

Comment:
move this inside the pagination function and execute concurrently for for each item in the list